satellite/satellitedb/overlaycache.go

```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
+		finalQuery := ""
+		finalArgs := []interface{}{}
+
+		// temporary arg slices are needed for each iteration, otherwise we'll end up with too many args appended
+		tempNewNodeArgs := []interface{}{}
+		tempNewNodeArgs = append(tempNewNodeArgs, newNodeArgs...)
+		tempReputableNodeArgs := []interface{}{}
+		tempReputableNodeArgs = append(tempReputableNodeArgs, reputableNodeArgs...)
+
+		if receivedNewNodeCount < newNodeCount {
+			newNodeQuery, moreNewNodeArgs := buildSelectionDistinct(ctx, criteria.ExcludedNetworks, newNodeCount, safeNewNodeQuery, true)
+			tempNewNodeArgs = append(tempNewNodeArgs, moreNewNodeArgs...)
+			finalArgs = append(finalArgs, tempNewNodeArgs...)
+			finalQuery = newNodeQuery + " UNION ALL "
+		}
+		reputableNodeQuery, moreReputableNodeArgs := buildSelectionDistinct(ctx, criteria.ExcludedNetworks, reputableNodeCount, safeReputableNodeQuery, false)
+		tempReputableNodeArgs = append(tempReputableNodeArgs, moreReputableNodeArgs...)
+
+		finalQuery += reputableNodeQuery
+		finalArgs = append(finalArgs, tempReputableNodeArgs...)
+
+		rows, err := cache.db.Query(ctx, cache.db.Rebind(finalQuery), finalArgs...)
 		if err != nil {
-			return nil, err
+			return nil, Error.Wrap(err)
+		}
+		defer func() { err = errs.Combine(err, rows.Close()) }()
+
+		var moreNodes []*overlay.SelectedNode
+		for rows.Next() {
+			var isNew bool
+			var node overlay.SelectedNode
+			node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+			err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort, &isNew)
+			if err != nil {
+				return nil, err
+			}
+			// checking for last net collision among reputable and new nodes since we can't check within the query
+			if _, ok := receivedNodeNetworks[node.LastNet]; ok {
+				continue
+			}
+			if isNew {
+				receivedNewNodeCount++
+				receivedNodeNetworks[node.LastNet] = struct{}{}
+			}
+			moreNodes = append(moreNodes, &node)
 		}
 		for _, n := range moreNodes {
 			nodes = append(nodes, n)
 			criteria.ExcludedIDs = append(criteria.ExcludedIDs, n.ID)
 			criteria.ExcludedNetworks = append(criteria.ExcludedNetworks, n.LastNet)
 		}
-		if len(nodes) == count {
+		if len(nodes) == totalCount {
 			break
 		}
 	}
-
 	return nodes, nil
 }
 
-func (cache *overlaycache) SelectNewStorageNodes(ctx context.Context, count int, criteria *overlay.NodeCriteria) (nodes []*overlay.SelectedNode, err error) {
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, isNewNodeQuery bool) (query string, args []interface{}, err error) {
 	defer mon.Task()(&ctx)(&err)
-
 	nodeType := int(pb.NodeType_STORAGE)
 
-	safeQuery := `
-		WHERE disqualified IS NULL
-		AND suspended IS NULL
-		AND exit_initiated_at IS NULL
-		AND type = ?
-		AND free_disk >= ?
+	// initiated with reputable node conditions
+	reputationCondition := `
```

Comment:
The condition query and the arguments are spread apart making this hard to follow.